File: src/progressStore.js
import { supabase } from "./supabaseClient";

/**
 * Math progress is PER KID. Locally the blob is scoped by the active kid
 * profile pointer (same scheme as engagementStore: `kidmath-progress:<kid>`,
 * bare key for anonymous play, first kid inherits the device blob once — copy,
 * never rename). In the cloud every row carries `kid_id`; rows with a null
 * kid_id are "household" rows merged from a device before any profile existed,
 * and the first kid to load a mode inherits them as a seed.
 *
 * The active-kid key is read directly (kidProfiles.js owns it) so the sync
 * loader the engine calls stays cheap and import-light.
 */
const PROGRESS_KEY = "kidmath-progress";
const ACTIVE_KID_KEY = "kidmath-active-kid"; // kidProfiles.js owns this key
const MIGRATED_KEY = "kidmath-progress-migrated"; // which kid inherited the device blob
const STARTING_LEVEL = 1;
const MAX_LEVEL = 10;

export function activeKidIdSync() {
  try {
    return localStorage.getItem(ACTIVE_KID_KEY) || null;
  } catch {
    return null;
  }
}

function storeKey(kidId) {
  return kidId ? `${PROGRESS_KEY}:${kidId}` : PROGRESS_KEY;
}

function clampLevel(level) {
  return Math.max(1, Math.min(MAX_LEVEL, level));
}

// --- localStorage helpers ---

function parseStore(raw) {
  try {
    const parsed = JSON.parse(raw);
    return parsed && typeof parsed === "object" ? parsed : {};
  } catch {
    return {};
  }
}

function readLocalStore(kidId = activeKidIdSync()) {
  const key = storeKey(kidId);
  try {
    const raw = localStorage.getItem(key);
    if (raw) return parseStore(raw);
    // First kid on this device inherits the anonymous blob — exactly once,
    // stamped even when there was nothing to inherit, so a later kid can
    // never claim it. Copy, never rename: renaming wipes progress.
    if (kidId && !localStorage.getItem(MIGRATED_KEY)) {
      localStorage.setItem(MIGRATED_KEY, kidId);
      const device = localStorage.getItem(PROGRESS_KEY);
      if (device) {
        localStorage.setItem(key, device);
        return parseStore(device);
      }
    }
    return {};
  } catch {
    return {};
  }
}

function writeLocalStore(store, kidId = activeKidIdSync()) {
  try {
    localStorage.setItem(storeKey(kidId), JSON.stringify(store));
  } catch {
    /* private mode / quota — the cloud copy is the durable one when signed in */
  }
}

function clearLocalStore(kidId) {
  try {
    localStorage.removeItem(storeKey(kidId));
  } catch {
    /* ignore */
  }
}

const MAX_PERSISTED_BANK_ITEMS = 200;
const MAX_PERSISTED_RECENT_IDS = 24;

function mergeBankItemStats(prev = {}, incoming = {}) {
  const merged = { ...prev };
  for (const [itemId, stats] of Object.entries(incoming)) {
    const base = merged[itemId] || {
      attempts: 0,
      firstTryCorrect: 0,
      correct: 0,
      totalResponseMs: 0,
      lastSeenAt: -1,
    };
    merged[itemId] = {
      attempts: (base.attempts ?? 0) + (stats.attempts ?? 0),
      firstTryCorrect: (base.firstTryCorrect ?? 0) + (stats.firstTryCorrect ?? 0),
      correct: (base.correct ?? 0) + (stats.correct ?? 0),
      totalResponseMs: (base.totalResponseMs ?? 0) + (stats.totalResponseMs ?? 0),
      lastSeenAt: Math.max(base.lastSeenAt ?? -1, stats.lastSeenAt ?? -1),
    };
  }
  const ids = Object.keys(merged);
  if (ids.length <= MAX_PERSISTED_BANK_ITEMS) return merged;
  const trimmed = ids
    .map((id) => [id, merged[id].lastSeenAt ?? -1])
    .sort((a, b) => b[1] - a[1])
    .slice(0, MAX_PERSISTED_BANK_ITEMS);
  return Object.fromEntries(trimmed.map(([id]) => [id, merged[id]]));
}

function loadLocal(mode, kidId) {
  const entry = readLocalStore(kidId)[mode];
  if (!entry) {
    return {
      level: STARTING_LEVEL,
      mistakeBank: [],
      totalSessions: 0,
      lifetimeStars: 0,
      bankItemStats: {},
      recentBankItemIds: [],
    };
  }
  return {
    level: clampLevel(entry.level ?? STARTING_LEVEL),
    mistakeBank: Array.isArray(entry.mistakeBank) ? entry.mistakeBank : [],
    totalSessions: entry.totalSessions ?? 0,
    lifetimeStars: entry.lifetimeStars ?? 0,
    bankItemStats: entry.bankItemStats && typeof entry.bankItemStats === "object" ? entry.bankItemStats : {},
    recentBankItemIds: Array.isArray(entry.recentBankItemIds) ? entry.recentBankItemIds : [],
  };
}

// `starsEarned` is the flight payout (§01 economy). Callers not yet on the
// Flight Report (and iOS until its port) omit it and keep the historical
// one-star-per-first-try formula.
function saveLocal(mode, { level, mistakeBank, firstTryCorrect, starsEarned, bankItemStats, recentBankItemIds }, kidId) {
  const store = readLocalStore(kidId);
  const prev = store[mode] || { totalSessions: 0, lifetimeStars: 0, bankItemStats: {} };
  store[mode] = {
    level: clampLevel(level),
    mistakeBank: (mistakeBank || []).slice(0, 20),
    totalSessions: (prev.totalSessions ?? 0) + 1,
    lifetimeStars: (prev.lifetimeStars ?? 0) + (starsEarned ?? firstTryCorrect ?? 0),
    bankItemStats: mergeBankItemStats(prev.bankItemStats || {}, bankItemStats || {}),
    recentBankItemIds: (recentBankItemIds || []).slice(-MAX_PERSISTED_RECENT_IDS),
  };
  writeLocalStore(store, kidId);
}

// --- Supabase helpers ---

async function getUser() {
  if (!supabase) return null;
  const { data: { session } } = await supabase.auth.getSession();
  return session?.user ?? null;
}

/** `.eq` cannot match NULL; household rows need `.is`. */
function forKid(query, kidId) {
  return kidId ? query.eq("kid_id", kidId) : query.is("kid_id", null);
}

async function loadCloudBankItemStats(userId, kidId, mode) {
  const { data, error } = await forKid(
    supabase
      .from("progress_item_stats")
      .select("item_id, attempts, first_try_correct, correct, total_response_ms, last_seen_at")
      .eq("user_id", userId)
      .eq("mode", mode),
    kidId
  );
  if (error || !Array.isArray(data)) return {};
  const stats = {};
  for (const row of data) {
    stats[row.item_id] = {
      attempts: row.attempts ?? 0,
      firstTryCorrect: row.first_try_correct ?? 0,
      correct: row.correct ?? 0,
      totalResponseMs: row.total_response_ms ?? 0,
      lastSeenAt: row.last_seen_at ? Date.parse(row.last_seen_at) : -1,
    };
  }
  return stats;
}

async function upsertCloudBankItemStats(userId, kidId, mode, bankItemStats) {
  if (!bankItemStats || Object.keys(bankItemStats).length === 0) return;
  const rows = Object.entries(bankItemStats).map(([itemId, stats]) => ({
    user_id: userId,
    kid_id: kidId || null,
    mode,
    item_id: itemId,
    attempts: stats.attempts ?? 0,
    first_try_correct: stats.firstTryCorrect ?? 0,
    correct: stats.correct ?? 0,
    total_response_ms: stats.totalResponseMs ?? 0,
    last_seen_at: new Date(
      Number.isFinite(stats.lastSeenAt) && stats.lastSeenAt > 0 ? stats.lastSeenAt : Date.now()
    ).toISOString(),
  }));
  await supabase
    .from("progress_item_stats")
    .upsert(rows, { onConflict: "user_id,kid_id,mode,item_id" });
}

async function fetchProgressRow(userId, kidId, mode) {
  const { data, error } = await forKid(
    supabase
      .from("progress")
      .select("level, mistake_bank, total_sessions, lifetime_stars, recent_bank_item_ids")
      .eq("user_id", userId)
      .eq("mode", mode),
    kidId
  ).limit(1);
  if (error || !Array.isArray(data)) return null;
  return data[0] || null;
}

/**
 * A kid's cloud progress for one mode. A kid with no row yet inherits the
 * household row (kid_id null — merged from a device before the profile
 * existed) as a seed; the first save writes their own row, so the inherit is
 * one-time by construction.
 */
async function loadCloud(userId, kidId, mode) {
  let [data, bankItemStats] = await Promise.all([
    fetchProgressRow(userId, kidId, mode),
    loadCloudBankItemStats(userId, kidId, mode),
  ]);
  if (!data && kidId) {
    [data, bankItemStats] = await Promise.all([
      fetchProgressRow(userId, null, mode),
      loadCloudBankItemStats(userId, null, mode),
    ]);
  }

  if (!data) {
    return {
      level: STARTING_LEVEL,
      mistakeBank: [],
      totalSessions: 0,
      lifetimeStars: 0,
      bankItemStats: bankItemStats || {},
      recentBankItemIds: [],
    };
  }
  return {
    level: clampLevel(data.level ?? STARTING_LEVEL),
    mistakeBank: Array.isArray(data.mistake_bank) ? data.mistake_bank : [],
    totalSessions: data.total_sessions ?? 0,
    lifetimeStars: data.lifetime_stars ?? 0,
    bankItemStats,
    recentBankItemIds: Array.isArray(data.recent_bank_item_ids) ? data.recent_bank_item_ids : [],
  };
}

async function saveCloud(userId, kidId, mode, { level, mistakeBank, firstTryCorrect, starsEarned, bankItemStats, recentBankItemIds }) {
  const existing = await loadCloud(userId, kidId, mode);
  const newTotalSessions = existing.totalSessions + 1;
  const newLifetimeStars = existing.lifetimeStars + (starsEarned ?? firstTryCorrect ?? 0);

  await Promise.all([
    supabase.from("progress").upsert(
      {
        user_id: userId,
        kid_id: kidId || null,
        mode,
        level: clampLevel(level),
        mistake_bank: (mistakeBank || []).slice(0, 20),
        total_sessions: newTotalSessions,
        lifetime_stars: newLifetimeStars,
        // Persisted since PR B so the no-repeat window survives a device switch.
        recent_bank_item_ids: (recentBankItemIds || []).slice(-MAX_PERSISTED_RECENT_IDS),
        updated_at: new Date().toISOString(),
      },
      { onConflict: "user_id,kid_id,mode" }
    ),
    upsertCloudBankItemStats(userId, kidId, mode, bankItemStats),
  ]);
}

// --- Merge localStorage into Supabase on first sign-in ---

/**
 * Merge this device's blob for the active kid (or the anonymous blob when no
 * profile is active) into that kid's cloud rows, then drop the local copy.
 * The anonymous blob is merged as household rows (kid_id null) and later
 * inherited by the first kid — see loadCloud.
 */
export async function mergeLocalToCloud(userId, kidId = activeKidIdSync()) {
  const store = readLocalStore(kidId);
  const modes = Object.keys(store);
  if (modes.length === 0) return;

  for (const mode of modes) {
    const local = store[mode];
    const cloud = await loadCloud(userId, kidId, mode);

    const merged = {
      level: Math.max(clampLevel(local.level ?? 1), cloud.level),
      mistake_bank: cloud.mistakeBank.length > 0 ? cloud.mistakeBank : (local.mistakeBank || []).slice(0, 20),
      total_sessions: cloud.totalSessions + (local.totalSessions ?? 0),
      lifetime_stars: cloud.lifetimeStars + (local.lifetimeStars ?? 0),
      updated_at: new Date().toISOString(),
    };

    await supabase.from("progress").upsert(
      { user_id: userId, kid_id: kidId || null, mode, ...merged },
      { onConflict: "user_id,kid_id,mode" }
    );

    // Migrate per-item stats: sum local counts into cloud counts.
    const mergedBankStats = mergeBankItemStats(cloud.bankItemStats || {}, local.bankItemStats || {});
    await upsertCloudBankItemStats(userId, kidId, mode, mergedBankStats);
  }

  clearLocalStore(kidId);
}

// --- Public API (same shape as the old mathEngine functions) ---

export async function loadProgress(mode, { kidId = activeKidIdSync() } = {}) {
  const user = await getUser();
  if (user) return loadCloud(user.id, kidId, mode);
  return loadLocal(mode, kidId);
}

export async function saveProgress(mode, data, { kidId = activeKidIdSync() } = {}) {
  const user = await getUser();
  if (user) {
    await saveCloud(user.id, kidId, mode, data);
  } else {
    saveLocal(mode, data, kidId);
  }
}

// Synchronous fallback for initial render before auth resolves (also the
// engine's injected progressLoader — keep it sync and cheap).
export function loadProgressSync(mode, kidId = activeKidIdSync()) {
  return loadLocal(mode, kidId);
}

/**
 * Every mode's progress in one read — the Grown-Ups panel's shape. Signed-in
 * families must get the cloud rows (the sign-in merge deletes the local blob,
 * so localStorage is empty for them); everyone else gets the local store.
 * `source` tells the UI which copy it is looking at.
 */
export async function loadProgressSummary({ kidId = activeKidIdSync() } = {}) {
  const user = await getUser();
  if (user) {
    let { data, error } = await forKid(
      supabase
        .from("progress")
        .select("mode, level, mistake_bank, total_sessions, lifetime_stars")
        .eq("user_id", user.id),
      kidId
    );
    // A kid with no rows of their own yet sees the household seed they would
    // inherit on first play — the same rule as loadCloud.
    if (!error && kidId && Array.isArray(data) && data.length === 0) {
      ({ data, error } = await supabase
        .from("progress")
        .select("mode, level, mistake_bank, total_sessions, lifetime_stars")
        .eq("user_id", user.id)
        .is("kid_id", null));
    }
    if (!error && Array.isArray(data)) {
      return {
        source: "cloud",
        byMode: Object.fromEntries(
          data.map((row) => [
            row.mode,
            {
              level: clampLevel(row.level ?? STARTING_LEVEL),
              mistakeBank: Array.isArray(row.mistake_bank) ? row.mistake_bank : [],
              totalSessions: row.total_sessions ?? 0,
              lifetimeStars: row.lifetime_stars ?? 0,
            },
          ])
        ),
      };
    }
  }
  return {
    source: "local",
    byMode: Object.fromEntries(
      Object.keys(readLocalStore(kidId)).map((mode) => [mode, loadLocal(mode, kidId)])
    ),
  };
}

File: ios/KidMath/Services/ProgressStore.swift
import Foundation

/// Per-mode saved progress — Swift mirror of src/progressStore.js.
///
/// Progress is PER KID: the local blob is scoped by the active kid pointer
/// (`kidmath-progress:<kid>`, bare key for anonymous play, first kid inherits
/// the device blob once — same scheme as EngagementStore) and every cloud row
/// carries `kid_id`. Rows with a null kid_id are "household" rows merged from
/// a device before any profile existed; the first kid to load a mode inherits
/// them as a seed.
///
/// Signed out: UserDefaults, same JSON shape as the web's localStorage entry.
/// Signed in: the shared `progress` + `progress_item_stats` tables, so a child
/// resumes at the same level on iPad and web. On first sign-in local progress
/// merges into the cloud and the local copy is cleared (mergeLocalToCloud).
///
/// `load` returns the exact `savedProgress` dictionary the engine expects in
/// `createSession(options: ["savedProgress": ...])`.
@MainActor
final class ProgressStore {

    nonisolated private static let localKey = "kidmath-progress"
    nonisolated private static let activeKidKey = "kidmath-active-kid" // KidProfilesService owns this key
    nonisolated private static let migratedKey = "kidmath-progress-migrated" // which kid inherited the device blob
    nonisolated private static let activeKidGradeKey = "kidmath-active-kid-grade" // KidProfilesService owns this key
    nonisolated private static let startingLevel = 1
    nonisolated private static let maxLevel = 10
    nonisolated private static let maxPersistedMistakes = 20
    nonisolated private static let maxPersistedBankItems = 200
    nonisolated private static let maxPersistedRecentIds = 24

    private let supabase: SupabaseService
    private let defaults: UserDefaults
    /// The shared engine, for the sign-in merge of skill state (set by
    /// AppModel once the engine is up; nil leaves skill fields unmerged).
    var engine: EngineBridge?

    init(supabase: SupabaseService = .shared, defaults: UserDefaults = .standard) {
        self.supabase = supabase
        self.defaults = defaults
    }

    /// Level a never-played mode opens at: 1 for anonymous play or an unknown
    /// grade, otherwise the kid's grade on the mode's ladder (GradeSeed) —
    /// same rule as the web's progressStore.startingLevelFor.
    nonisolated static func startingLevel(mode: String, defaults: UserDefaults = .standard) -> Int {
        guard defaults.string(forKey: activeKidKey).map({ !$0.isEmpty }) ?? false else { return startingLevel }
        return GradeSeed.startingLevel(mode: mode, grade: defaults.string(forKey: activeKidGradeKey))
    }

    nonisolated static func blankProgress(mode: String? = nil) -> [String: Any] {
        [
            "level": mode.map { startingLevel(mode: $0) } ?? startingLevel,
            "mistakeBank": [[String: Any]](),
            "totalSessions": 0,
            "lifetimeStars": 0,
            "bankItemStats": [String: Any](),
            "recentBankItemIds": [String](),
        ]
    }

    // MARK: - Public API (same shape as the web's loadProgress/saveProgress)

    /// The active kid profile id, or nil for anonymous play.
    var activeKidId: UUID? {
        guard let raw = defaults.string(forKey: Self.activeKidKey), !raw.isEmpty else { return nil }
        return UUID(uuidString: raw)
    }

    func load(mode: String) async -> [String: Any] {
        if let userId = supabase.userId {
            return await loadCloud(userId: userId, kidId: activeKidId, mode: mode)
        }
        return loadLocal(mode: mode)
    }

    /// `data` comes from a finished engine session snapshot: level,
    /// mistakeBank, firstTryCorrect, bankItemStats, recentBankItemIds.
    func save(mode: String, data: [String: Any]) async {
        if let userId = supabase.userId {
            await saveCloud(userId: userId, kidId: activeKidId, mode: mode, data: data)
        } else {
            saveLocal(mode: mode, data: data)
        }
    }

    /// First sign-in: fold this device's progress for the active kid (or the
    /// anonymous blob when no profile is active — merged as household rows)
    /// into the cloud, then clear the local copy.
    func mergeLocalToCloud(userId: UUID) async {
        let kidId = activeKidId
        let store = readLocalStore()
        guard !store.isEmpty else { return }
        for (mode, local) in store {
            let cloud = await loadCloud(userId: userId, kidId: kidId, mode: mode)
            let localMistakes = (local["mistakeBank"] as? [[String: Any]] ?? [])
            let cloudMistakes = (cloud["mistakeBank"] as? [[String: Any]] ?? [])
            var merged: [String: Any] = [
                "level": max(
                    Self.clampLevel(Self.int(local["level"], default: 1), mode: mode),
                    Self.int(cloud["level"], default: Self.startingLevel)
                ),
                "mistake_bank": cloudMistakes.isEmpty ? Array(localMistakes.prefix(Self.maxPersistedMistakes)) : cloudMistakes,
                "total_sessions": Self.int(cloud["totalSessions"]) + Self.int(local["totalSessions"]),
                "lifetime_stars": Self.int(cloud["lifetimeStars"]) + Self.int(local["lifetimeStars"]),
            ]
            // Nothing earned is lost: the shared mergeTopicState (skills/topicState.js).
            if let engine {
                let topic = engine.mergeTopicState(cloud: Self.skillFields(cloud), local: Self.skillFields(local))
                merged.merge(Self.skillColumns(Self.skillPatch(topic))) { _, new in new }
            }
            let mergedStats = Self.mergeBankItemStats(
                cloud["bankItemStats"] as? [String: [String: Any]] ?? [:],
                incoming: local["bankItemStats"] as? [String: [String: Any]] ?? [:]
            )
            do {
                try await supabase.upsertProgress(userId: userId, kidId: kidId, mode: mode, row: merged)
                try await supabase.upsertBankItemStats(userId: userId, kidId: kidId, mode: mode, stats: mergedStats)
            } catch {
                return // keep local copy; retry on a later sign-in
            }
        }
        defaults.removeObject(forKey: storeKey)
    }

    // MARK: - Local (UserDefaults, same JSON shape as web localStorage; per-kid keys)

    private var storeKey: String {
        if let kid = defaults.string(forKey: Self.activeKidKey), !kid.isEmpty {
            return "\(Self.localKey):\(kid)"
        }
        return Self.localKey
    }

    private func readBlob(_ key: String) -> [String: [String: Any]]? {
        guard let data = defaults.data(forKey: key) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: [String: Any]]
    }

    private func readLocalStore() -> [String: [String: Any]] {
        let key = storeKey
        if let store = readBlob(key) { return store }
        // First kid on this device inherits the anonymous blob — exactly once,
        // stamped even when there was nothing to inherit. Copy, never rename.
        if key != Self.localKey, defaults.string(forKey: Self.migratedKey) == nil {
            defaults.set(String(key.dropFirst(Self.localKey.count + 1)), forKey: Self.migratedKey)
            if let device = readBlob(Self.localKey) {
                writeLocalStore(device)
                return device
            }
        }
        return [:]
    }

    private func writeLocalStore(_ store: [String: [String: Any]]) {
        if let data = try? JSONSerialization.data(withJSONObject: store) {
            defaults.set(data, forKey: storeKey)
        }
    }

    func loadLocal(mode: String) -> [String: Any] {
        guard let entry = readLocalStore()[mode] else { return Self.blankProgress(mode: mode) }
        return [
            "level": Self.clampLevel(Self.int(entry["level"], default: Self.startingLevel), mode: mode),
            "mistakeBank": entry["mistakeBank"] as? [[String: Any]] ?? [],
            "totalSessions": Self.int(entry["totalSessions"]),
            "lifetimeStars": Self.int(entry["lifetimeStars"]),
            "bankItemStats": entry["bankItemStats"] as? [String: Any] ?? [:],
            "recentBankItemIds": entry["recentBankItemIds"] as? [String] ?? [],
        ].merging(Self.skillFields(entry)) { _, new in new }
    }

    func saveLocal(mode: String, data: [String: Any]) {
        var store = readLocalStore()
        let previous = store[mode] ?? [:]
        let mistakes = (data["mistakeBank"] as? [[String: Any]] ?? []).prefix(Self.maxPersistedMistakes)
        let recent = (data["recentBankItemIds"] as? [String] ?? []).suffix(Self.maxPersistedRecentIds)
        store[mode] = [
            "level": Self.clampLevel(Self.int(data["level"], default: Self.startingLevel), mode: mode),
            "mistakeBank": Array(mistakes),
            "totalSessions": Self.int(previous["totalSessions"]) + 1,
            "lifetimeStars": Self.int(previous["lifetimeStars"]) + Self.starsEarned(from: data),
            "bankItemStats": Self.mergeBankItemStats(
                previous["bankItemStats"] as? [String: [String: Any]] ?? [:],
                incoming: data["bankItemStats"] as? [String: [String: Any]] ?? [:]
            ),
            "recentBankItemIds": Array(recent),
        ]
        .merging(Self.skillFields(previous)) { _, new in new }
        .merging(Self.skillPatch(data)) { _, new in new }
        writeLocalStore(store)
    }

    // MARK: - Play by skill (the topic's grade pointer + mastery ride on the progress row)

    /// Change a topic's skill state WITHOUT counting a session: a grown-up
    /// opening a grade or pinning a skill, the lazy grade migration, a
    /// grade-up. Never touches level, stars or the session count (mirror of
    /// the web's saveTopicState).
    func saveTopicState(mode: String, patch: [String: Any]) async {
        let fields = Self.skillPatch(patch)
        guard !fields.isEmpty else { return }
        if let userId = supabase.userId {
            let kidId = activeKidId
            // A row must exist to carry the fields; a never-played topic gets
            // one at the level it would have opened at — which is not a change.
            let existing = try? await supabase.fetchProgressRow(userId: userId, kidId: kidId, mode: mode)
            var row = Self.skillColumns(fields)
            row["level"] = existing.map { Self.int($0["level"], default: Self.startingLevel) } ?? Self.startingLevel(mode: mode, defaults: defaults)
            try? await supabase.upsertProgress(userId: userId, kidId: kidId, mode: mode, row: row)
            return
        }
        var store = readLocalStore()
        var entry = store[mode] ?? Self.blankProgress(mode: mode)
        entry.merge(fields) { _, new in new }
        store[mode] = entry
        writeLocalStore(store)
    }

    // MARK: - Cloud

    /// A kid with no row yet inherits the household row (kid_id null) as a
    /// seed; their first save writes their own row (web parity).
    private func loadCloud(userId: UUID, kidId: UUID?, mode: String) async -> [String: Any] {
        var stats = (try? await supabase.fetchBankItemStats(userId: userId, kidId: kidId, mode: mode)) ?? [:]
        var row = try? await supabase.fetchProgressRow(userId: userId, kidId: kidId, mode: mode)
        if row == nil, kidId != nil {
            row = try? await supabase.fetchProgressRow(userId: userId, kidId: nil, mode: mode)
            stats = (try? await supabase.fetchBankItemStats(userId: userId, kidId: nil, mode: mode)) ?? [:]
        }
        guard let row else {
            var blank = Self.blankProgress(mode: mode)
            blank["bankItemStats"] = stats
            return blank
        }
        return [
            "level": Self.clampLevel(Self.int(row["level"], default: Self.startingLevel), mode: mode),
            "mistakeBank": row["mistake_bank"] as? [[String: Any]] ?? [],
            "totalSessions": Self.int(row["total_sessions"]),
            "lifetimeStars": Self.int(row["lifetime_stars"]),
            "bankItemStats": stats,
            // Persisted since PR B so the no-repeat window survives a device switch.
            "recentBankItemIds": row["recent_bank_item_ids"] as? [String] ?? [],
        ].merging(Self.skillFields(fromRow: row)) { _, new in new }
    }

    private func saveCloud(userId: UUID, kidId: UUID?, mode: String, data: [String: Any]) async {
        let existing = await loadCloud(userId: userId, kidId: kidId, mode: mode)
        var row: [String: Any] = [
            "level": Self.clampLevel(Self.int(data["level"], default: Self.startingLevel), mode: mode),
            "mistake_bank": Array((data["mistakeBank"] as? [[String: Any]] ?? []).prefix(Self.maxPersistedMistakes)),
            "total_sessions": Self.int(existing["totalSessions"]) + 1,
            "lifetime_stars": Self.int(existing["lifetimeStars"]) + Self.starsEarned(from: data),
            "recent_bank_item_ids": Array((data["recentBankItemIds"] as? [String] ?? []).suffix(Self.maxPersistedRecentIds)),
        ]
        // Only a skill session carries these; a ladder session leaves them alone.
        row.merge(Self.skillColumns(Self.skillPatch(data))) { _, new in new }
        try? await supabase.upsertProgress(userId: userId, kidId: kidId, mode: mode, row: row)
        try? await supabase.upsertBankItemStats(
            userId: userId,
            kidId: kidId,
            mode: mode,
            stats: data["bankItemStats"] as? [String: [String: Any]] ?? [:]
        )
    }

    // MARK: - Pure helpers (mirrors of the web functions, unit-tested)

    nonisolated static let skillKeys = ["grade", "gradeUnlocked", "pinnedSkillId", "skillMastery"]
    nonisolated private static let skillColumnNames = [
        "grade": "grade", "gradeUnlocked": "grade_unlocked", "pinnedSkillId": "pinned_skill_id", "skillMastery": "skill_mastery",
    ]

    /// The skill fields a stored entry carries — absent ones are simply
    /// absent (the shared resolveTopic fills the gaps, never the store).
    nonisolated static func skillFields(_ entry: [String: Any]) -> [String: Any] {
        var out: [String: Any] = [:]
        for key in skillKeys {
            if let value = entry[key], !(value is NSNull) { out[key] = value }
        }
        return out
    }

    nonisolated static func skillFields(fromRow row: [String: Any]) -> [String: Any] {
        var entry: [String: Any] = [:]
        for (key, column) in skillColumnNames { entry[key] = row[column] }
        return skillFields(entry)
    }

    /// Skill fields present in a save payload — absent keys leave the stored
    /// value alone; an explicit NSNull clears (un-pinning a skill).
    nonisolated static func skillPatch(_ data: [String: Any]) -> [String: Any] {
        var out: [String: Any] = [:]
        for key in skillKeys {
            if let value = data[key] { out[key] = value }
        }
        return out
    }

    nonisolated static func skillColumns(_ patch: [String: Any]) -> [String: Any] {
        var out: [String: Any] = [:]
        for (key, value) in patch {
            if let column = skillColumnNames[key] { out[column] = value }
        }
        return out
    }

    /// Numbers arrive as JSC doubles, JSON NSNumbers, or native Int/Double
    /// literals; only NSNumber bridges all three, so never `as? Int` directly.
    nonisolated static func int(_ value: Any?, default fallback: Int = 0) -> Int {
        (value as? NSNumber)?.intValue ?? fallback
    }

    nonisolated static func double(_ value: Any?, default fallback: Double = 0) -> Double {
        (value as? NSNumber)?.doubleValue ?? fallback
    }

    nonisolated static func clampLevel(_ level: Int, mode: String? = nil) -> Int {
        let cap = mode.map { GradeSeed.maxLevel(mode: $0) } ?? maxLevel
        return min(cap, max(1, level))
    }

    /// §01: `starsEarned` (the flight payout) wins when present — including an
    /// explicit 0 — and callers not yet on the Flight Report fall back to the
    /// historical one-star-per-first-try formula. Mirror of the web's
    /// `starsEarned ?? firstTryCorrect` in progressStore.js.
    nonisolated static func starsEarned(from data: [String: Any]) -> Int {
        if let explicit = data["starsEarned"] { return int(explicit) }
        return int(data["firstTryCorrect"])
    }

    /// Sum incoming per-item counters into the previous ones, keep the most
    /// recently seen 200 items (mirror of mergeBankItemStats in progressStore.js).
    nonisolated static func mergeBankItemStats(
        _ previous: [String: [String: Any]],
        incoming: [String: [String: Any]]
    ) -> [String: [String: Any]] {
        var merged = previous
        for (itemId, stats) in incoming {
            let base = merged[itemId] ?? [:]
            merged[itemId] = [
                "attempts": Self.int(base["attempts"]) + Self.int(stats["attempts"]),
                "firstTryCorrect": Self.int(base["firstTryCorrect"]) + Self.int(stats["firstTryCorrect"]),
                "correct": Self.int(base["correct"]) + Self.int(stats["correct"]),
                "totalResponseMs": Self.int(base["totalResponseMs"]) + Self.int(stats["totalResponseMs"]),
                "lastSeenAt": max(Self.double(base["lastSeenAt"], default: -1), Self.double(stats["lastSeenAt"], default: -1)),
            ]
        }
        guard merged.count > maxPersistedBankItems else { return merged }
        let kept = merged
            .sorted { Self.double($0.value["lastSeenAt"], default: -1) > Self.double($1.value["lastSeenAt"], default: -1) }
            .prefix(maxPersistedBankItems)
        return Dictionary(uniqueKeysWithValues: kept.map { ($0.key, $0.value) })
    }
}
